crates/fuel-core/src/query/message.rs

```diff
@@ -167,52 +152,107 @@ pub fn message_proof<T: MessageProofData + ?Sized>(
                 recipient,
                 nonce,
                 amount,
-                data: message_data,
+                data,
                 ..
             } if r.message_id() == Some(message_id) => {
-                Some((sender, recipient, nonce, amount, message_data))
+                Some((sender, recipient, nonce, amount, data))
             }
             _ => None,
         });
 
-    let (sender, recipient, nonce, amount, message_data) = match receipt {
+    let (sender, recipient, nonce, amount, data) = match receipt {
         Some(r) => r,
         None => return Ok(None),
     };
 
     // Get the block id from the transaction status if it's ready.
-    let block_id = match data
+    let message_block_id = match database
         .transaction_status(&transaction_id)
         .into_api_result::<TransactionStatus, StorageError>()?
     {
         Some(TransactionStatus::Success { block_id, .. }) => block_id,
         _ => return Ok(None),
     };
 
-    // Get the message ids in the same order as the transactions.
-    let leaves: Vec<Vec<Receipt>> = data
-        .transactions_on_block(&block_id)?
-        .into_iter()
-        .map(|id| data.receipts(&id))
+    // Get the message fuel block header.
+    let (message_block_header, message_block_txs) = match database
+        .block(&message_block_id)
+        .into_api_result::<CompressedBlock, StorageError>()?
+    {
+        Some(t) => t.into_inner(),
+        None => return Ok(None),
+    };
+
+    let (root, message_proof) =
+        match message_receipts_proof(database, message_id, &message_block_txs)? {
+            Some(proof) => proof,
+            None => return Ok(None),
+        };
+
+    if message_block_header.message_receipt_root != root {
```

Comment:
Do we know if this can actually happen? Maybe we can omit this if we have good test coverage to make sure our proofs are generated correctly and that our block production works correctly. Then, we also don't need the root here.